File: client/dexterity/codegen/instruments/types/derivative_metadata.py
# LOCK-BEGIN[imports]: DON'T MODIFY
from dexterity.codegen.dex.types.fractional import Fractional
from dexterity.codegen.instruments.types.account_tag import AccountTag
from dexterity.codegen.instruments.types.expiration_status import ExpirationStatus
from dexterity.codegen.instruments.types.instrument_type import InstrumentType
from dexterity.codegen.instruments.types.oracle_type import OracleType
from podite import (
    U64,
    pod,
)
from solana.publickey import PublicKey
from solmate.dtypes import UnixTimestamp

# LOCK-END


# LOCK-BEGIN[class(DerivativeMetadata)]: DON'T MODIFY
@pod
class DerivativeMetadata:
    tag: AccountTag
    expired: ExpirationStatus
    oracle_type: OracleType
    instrument_type: InstrumentType
    bump: U64
    strike: Fractional
    initialization_time: UnixTimestamp
    full_funding_period: UnixTimestamp
    minimum_funding_period: UnixTimestamp
    price_oracle: PublicKey
    market_product_group: PublicKey
    close_authority: PublicKey
    clock: PublicKey
    last_funding_time: UnixTimestamp
    # LOCK-END

    @classmethod
    def to_bytes(cls, obj, **kwargs):
        return cls.pack(obj, converter="bytes", **kwargs)

    @classmethod
    def from_bytes(cls, raw, **kwargs):
        return cls.unpack(raw, converter="bytes", **kwargs)

File: programs/instruments/src/state/derivative_metadata.rs
use crate::state::enums::{AccountTag, ExpirationStatus, InstrumentType, OracleType};
use anchor_lang::prelude::*;
use dex::utils::numeric::Fractional;
use solana_program::{clock::UnixTimestamp, program_error::ProgramError, pubkey::Pubkey};

#[account(zero_copy)]
pub struct DerivativeMetadata {
    pub tag: AccountTag,
    pub expired: ExpirationStatus,
    pub oracle_type: OracleType,
    pub instrument_type: InstrumentType,
    pub bump: u64,
    pub strike: Fractional,
    pub initialization_time: UnixTimestamp,
    pub full_funding_period: UnixTimestamp,
    pub minimum_funding_period: UnixTimestamp,
    pub price_oracle: Pubkey,
    pub market_product_group: Pubkey,
    pub close_authority: Pubkey,
    pub clock: Pubkey,
    pub last_funding_time: UnixTimestamp,
}

impl DerivativeMetadata {
    pub fn get_key(&self, program_id: &Pubkey) -> std::result::Result<Pubkey, ProgramError> {
        let seeds = &[
            b"derivative",
            self.price_oracle.as_ref(),
            self.market_product_group.as_ref(),
            &(self.instrument_type as u64).to_le_bytes(),
            &self.strike.m.to_le_bytes(),
            &self.strike.exp.to_le_bytes(),
            &self.initialization_time.to_le_bytes(),
            &self.full_funding_period.to_le_bytes(),
            &self.minimum_funding_period.to_le_bytes(),
            &[self.bump as u8],
        ];
        Ok(Pubkey::create_program_address(seeds, program_id)?)
    }

    pub fn is_initialized(&self) -> bool {
        self.tag == AccountTag::DerivativeMetadata && self.expired == ExpirationStatus::Active
    }

    pub fn expired(&self) -> bool {
        self.expired == ExpirationStatus::Expired
    }
}
